sys/fork.c

```diff
@@ -0,0 +1,64 @@
+#include <fork.h>
+#include <thread.h>
+#include <sync.h>
+#include <filedesc.h>
+#include <sched.h>
+#include <stdc.h>
+
+int do_fork() {
+  thread_t *td = thread_self();
+
+  /* Cannot fork non-user threads. */
+  assert(td->td_uspace);
+
+  thread_t *newtd = thread_create(td->td_name, NULL, NULL);
+
+  /* Prevent preemption, so that the copied thread won't run while we're copying
+     it. */
+  critical_enter();
```

Comment:
Actually, I think we might drop the critical section here altogether. The nature of `fork` is such that copying happens within the context of the source thread, so its state will not change while `fork` is performed!

Copying address space is different, because it may be already shared between multiple threads, so some other thread could modify address space layout while it is being copied. Using a mutex to lock access to the vm map structure will not suffice - it would protect structure integrity, but the actual contents, the data in address space, could still be modified.

We would need to pick all threads that share the copied address space, and ask the scheduler not to run them until we are done. This would have to be done under a critical section, but would take little time. (Same goes for unblocking these threads when we are done copying). Would that be feasible?